### PR TITLE
Improve overloaded function call diagnostics

### PR DESCRIPTION
**Pull request by seanyoung**
From Solidity 0.6 onwards, when an overloaded function/event resolves to multiple candidates, this is an error. In earlier versions, the first result is used. So, use the `pragma solidity` version to decide whether to error or not.

Fixes: https://github.com/hyperledger/solang/issues/1534
Fixes: https://github.com/hyperledger/solang/issues/707

### CODE REVIEW

File: src/sema/expression/constructor.rs [hunk 1]
```diff
@@ -108,84 +108,99 @@ pub fn match_constructor_to_args(
...
+
+            Err(())
         }
+        1 => Ok(resolved_calls.remove(0)),
         _ => {
-            diagnostics.push(Diagnostic::error(
+            diagnostics.push(Diagnostic::error_with_notes(
                 *loc,
-                "cannot find overloaded constructor which matches signature".to_string(),
+                "constructor can be resolved to multiple functions".into(),
```
Comment (LucasSte) [inline src/sema/expression/constructor.rs:188]: There is no test for this.

File: src/sema/expression/constructor.rs [hunk 2]
```diff
@@ -344,86 +386,107 @@ pub fn constructor_named_args(
...
+            // check if arguments can be implicitly casted
+            for i in 0..params_len {
+                let param = ns.functions[*function_no].params[i].clone();
+
+                let arg = match arguments.get(param.name_as_str()) {
+                    Some(a) => a,
+                    None => {
+                        candidate_diagnostics.push(Diagnostic::cast_error_with_note(
+                            *loc,
+                            format!("missing argument '{}' to constructor", param.name_as_str()),
```
Comment (LucasSte) [inline src/sema/expression/constructor.rs:399]: No test for this message.

File: src/sema/expression/constructor.rs [hunk 3]
```diff
@@ -344,86 +386,107 @@ pub fn constructor_named_args(
...
         }),
-        0 | 1 => {
-            diagnostics.extend(errors);
+        0 => {
+            diagnostics.extend(call_diagnostics);
+
+            if function_nos.is_empty() {
+                diagnostics.push(Diagnostic::error(
+                    *loc,
+                    "cannot find matching constructor".into(),
```
Comment (LucasSte) [inline src/sema/expression/constructor.rs:464]: No test.

File: src/sema/expression/function_call.rs [hunk 4]
```diff
@@ -2709,67 +2817,91 @@ fn contract_call_named_args(
...
-                    ns.contracts[external_contract_no].id, func_name.name
-                ),
-            ));
+            diagnostics.extend(call_diagnostics);
+
+            if name_matches.is_empty() {
+                diagnostics.push(Diagnostic::error(
+                    *loc,
+                    format!(
+                        "contract '{}' does not have function '{}'",
```
Comment (LucasSte) [inline src/sema/expression/function_call.rs:2877]: Code path not tested.

File: src/sema/expression/function_call.rs [hunk 5]
```diff
@@ -2805,106 +2937,140 @@ fn contract_call_pos_args(
...
-        }
 
-        if matches {
-            let resolved_call = contract_call_match(
+            if candidate_diagnostics.any_errors() {
+                if name_matches.len() != 1 {
+                    // will be de-duped
+                    candidate_diagnostics.push(Diagnostic::error(
+                        *loc,
+                        "cannot find overloaded function which matches signature".into(),
```
Comment (LucasSte) [inline src/sema/expression/function_call.rs:2986]: Code path not tested.

File: src/sema/expression/function_call.rs [hunk 6]
```diff
@@ -2805,106 +2937,140 @@ fn contract_call_pos_args(
...
-        diagnostics.extend(errors);
-    } else if name_matches.len() != 1 {
-        diagnostics.push(Diagnostic::error(
-            *loc,
-            "cannot find overloaded function which matches signature".to_string(),
-        ));
-    }
+            diagnostics.push(Diagnostic::error_with_notes(
+                *loc,
+                "function call can be resolved to multiple functions".into(),
```
Comment (LucasSte) [inline src/sema/expression/function_call.rs:3057]: This error message is not tested.

File: src/sema/pragma.rs [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+                    res.push(v);
+                }
+
+                ns.pragmas.push(ast::Pragma::SolidityVersion {
+                    loc: *loc,
+                    versions: res,
+                });
+            }
+        }
+        pt::PragmaDirective::Identifier { .. } => (),
```
Comment (LucasSte) [inline src/sema/pragma.rs:51]: Is this case supposed to be an error?

**Comment by seanyoung** [comment]
@LucasSte Thanks for your "there is no test for this" comments. I am aware of this. Not exactly a useful review.

File: src/sema/pragma.rs (hunk 7 above)
Comment (seanyoung) [reply src/sema/pragma.rs:51]: no

**Review by xermicus — COMMENTED** [review]
LGTM overall. How / where are we going to document this?

File: src/sema/pragma.rs [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+            *loc,
+            "experimental solidity features are not supported".to_string(),
+        ));
+    } else if name == "abicoder" && value == "v2" {
+        ns.diagnostics.push(ast::Diagnostic::debug(
+            *loc,
+            "pragma 'abicoder' with value 'v2' is ignored".to_string(),
+        ));
+    } else {
+        ns.diagnostics.push(ast::Diagnostic::warning(
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply src/sema/pragma.rs]: oh nice, that reduces the failed ethereum solidity tests even more :rocket:

**Comment by seanyoung** [comment]
> How / where are we going to document this?

I had not thought about this. Do you have any suggestions?

**Comment by xermicus** [comment]
> > How / where are we going to document this?
> 
> I had not thought about this. Do you have any suggestions?

I guess it depends on the general direction we want to go with this. Currently, we just say in the docs that we ignore the version pragma, which is kind of no longer true after this PR.

**Review by xermicus — COMMENTED** [review]
If I got the following source:

```solidity
// a.sol
pragma solidity ^0.7;

import "b.sol";

contract C {
        function c() public payable {
                require(msg.value > SERVICE_FEE); 
        }
}
```

```solidity
// b.sol
pragma solidity ^0.8;

uint128 constant SERVICE_FEE = 123;
```

Shouldn't Solang then refuse to compile that?

**Comment by xermicus** [comment]
Maybe it's a sensible strategy to have each Solang release being compatible with a specific  `solc` release. I think that would clarify a lot of things.

**Comment by seanyoung** [comment]
> I guess it depends on the general direction we want to go with this. Currently, we just say in the docs that we ignore the version pragma, which is kind of no longer true after this PR.

There are some differences between versions, but nothing huge. The idea here is that we switch behaviour based on the solidity pragma, so that - ultimately - we can compile any Solidity. A single solidity compiler for all versions.

**Comment by seanyoung** [comment]
> If I got the following source:
> 
> ```solidity
> // a.sol
> pragma solidity ^0.7;
> 
> import "b.sol";
> 
> contract C {
>         function c() public payable {
>                 require(msg.value > SERVICE_FEE); 
>         }
> }
> ```
> 
> ```solidity
> // b.sol
> pragma solidity ^0.8;
> 
> uint128 constant SERVICE_FEE = 123;
> ```
> 
> Shouldn't Solang then refuse to compile that?

No, why? Just because solc only support one version does not mean we have to be limited to one version too.

**Comment by seanyoung** [comment]
> Maybe it's a sensible strategy to have each Solang release being compatible with a specific `solc` release. I think that would clarify a lot of things.

That would mean supporting lots of different versions of solang, one for each version of Solidity. That's too much work.

File: src/sema/expression/function_call.rs [hunk 9]
```diff
@@ -2805,106 +2937,141 @@ fn contract_call_pos_args(
...
 
-    let mut errors = Diagnostics::default();
+    let mut call_diagnostics = Diagnostics::default();
+    let mut resolved_calls = Vec::new();
+
+    // try to resolve the arguments, give up if there are any errors
+    if args.iter().fold(false, |acc, arg| {
+        acc | expression(arg, context, ns, symtable, diagnostics, ResolveTo::Unknown).is_err()
+    }) {
+        return Err(());
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply src/sema/expression/function_call.rs:2947]: The test coverage is already much better than what it was. I have added a test case.

**Comment by xermicus** [comment]
I agree that'd be a lot of work.

What about a warning if there are conflicting version pragma found in the source. This way, the user at least gets warned that he is trying to compile "incompatible" sources.

File: tests/contract_testcases/evm/builtin_overloaded.sol [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+	}
+
+	function test2() public {
+		bs.readInt8(1, "foo");
+	}
+}
+
+// ---- Expect: diagnostics ----
+// error: 3:20-34: cannot call function in constant expression
+// error: 6:15-20: implicit conversion to uint32 from bytes3 not allowed
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply tests/contract_testcases/evm/builtin_overloaded.sol:16]: All our conversion errors and warning say "to X from Y". I am not changing this in this PR.

**Comment by seanyoung** [comment]
> I agree that'd be a lot of work.
> 
> What about a warning if there are conflicting version pragma found in the source. This way, the user at least gets warned that he is trying to compile "incompatible" sources.

The problem is, how do you determine that two pragmas are conflicting? We might have in one file:
`pragma solidity 0.5 - 0.5.16;`
and another:
`pragma solidity <=0.6;`

In the solidity pragma parsing, all we have right now in solang is what the highest supported Solidity version. In this case we have `0.5.16`  and `0.6`. Is that a conflict? If you were to compile with solc 0.5.15 then it works fine, but if you compile with solc 0.6.1 then it is not fine.

I think this is something for another PR.

**Comment by xermicus** [comment]
By conflicting I meant there is no suitable compiler, (e.g. `pragma solidity ^0.5.16;` and `pragma solidity ^0.6.16`)

File: tests/contract_testcases/evm/overloaded.sol [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+// error: 6:3-23: cannot find overloaded function which matches signature
+// 	note 2:2-29: candidate function
+// error: 6:3-23: cannot find overloaded function which matches signature
+// 	note 3:2-39: candidate function
+// error: 6:3-23: cannot find overloaded function which matches signature
+// 	note 4:2-39: candidate function
+// error: 6:3-23: function cannot be called with named arguments as 1 of its parameters do not have names
+// 	note 2:2-26: definition of foo
+// 	note 2:2-29: candidate function
+// error: 6:11-15: conversion from bool to int256 not possible
```
Comment (LucasSte) [inline tests/contract_testcases/evm/overloaded.sol:56]: > All our conversion errors and warning say "to X from Y".

@seanyoung This is not true.

**Review by xermicus — APPROVED** [review]
> I don't understand why that is relevant to this PR. Is there some sort of regression?

No, it's not a regression. I still think it's relevant. As the PR description says:

> From Solidity 0.6 onwards, when an overloaded function/event resolves to multiple candidates, this is an error. In earlier versions, the first result is used. So, use the pragma solidity version to decide whether to error or not.

Which means Solang will _silently_ compile identical source codes differently, depending on the version pragma. I just thought it would be nice to inform our users when this happens. It might not be what they expected.  I _think_ it would be nice to emit some warnings if the compiler

- Encounters `version` pragmas that are mutually exclusive (meaning there would be no `solc` version that can compile this)
- Is compiling the same code differently due to different version pragmas

File: tests/contract_testcases/evm/overloaded.sol (hunk 11 above)
Comment (seanyoung) [reply tests/contract_testcases/evm/overloaded.sol:56]: @LucasSte as always, pedantic and not constructive.

**Comment by xermicus** [comment]
Sure, can be done later.

> warning: in later versions of solidity constructor visibility is ignored but if you remove it your contract will no longer build due to the pragma solidity 0.5;

No, that wouldn't be useful. What I meant was:

```
pragma solidity 0.6;

contract Fibonacci {
    constructor() {}
}
```

This should produce a warning like follows:

`Warning: Solidity 0.6 requires constructors to have a visibility.`

Which would remind the user that he is doing something that was not intended in Ethereum Solidity but works anyways in Solang.

**Comment by seanyoung** [comment]
It does make sense. Also, inth

After some more testing I've re-jigged the code to match the actual behaviour better, and also better error messages which shows the actual change in behaviour between versions which helps the developer.

**Comment by seanyoung** [comment]
let me know what you think

**Comment by xermicus** [comment]
Great! 

> make it clear to the developer what has changed and how it affects their code - in case there are any potential pitfalls.

Yeah this is exactly what I meant :)

**Pull request merged**